kernel/src/fs/path/dentry.rs

```diff
@@ -375,71 +389,94 @@ enum DentryOptions {
     Leaf((String, Arc<Dentry_>)),
 }
 
-struct Children {
-    dentries: HashMap<String, Arc<Dentry_>>,
+/// Manages child dentries, including both valid and negative entries.
+///
+/// A _negative_ dentry reflects a failed filename lookup, saving potential
+/// repeated and costly lookups in the future.
+// TODO: Address the issue of negative dentry bloating.
+struct DentryChildren {
+    dentries: HashMap<String, Option<Arc<Dentry_>>>,
 }
 
-impl Children {
+impl DentryChildren {
+    /// Creates an empty dentry cache.
     pub fn new() -> Self {
         Self {
             dentries: HashMap::new(),
         }
     }
 
-    pub fn len(&self) -> usize {
-        self.dentries.len()
+    /// Checks if a valid dentry with the given name exists.
+    pub fn contains_valid(&self, name: &str) -> bool {
+        self.dentries.get(name).is_some_and(|child| child.is_some())
     }
 
-    pub fn contains(&self, name: &str) -> bool {
-        self.dentries.contains_key(name)
+    /// Checks if a negative dentry with the given name exists.
+    pub fn contains_negative(&self, name: &str) -> bool {
+        self.dentries.get(name).is_some_and(|child| child.is_none())
     }
 
-    pub fn find(&self, name: &str) -> Option<Arc<Dentry_>> {
-        self.dentries.get(name).cloned()
+    /// Finds a dentry by name. Returns error for negative entries.
+    pub fn find(&self, name: &str) -> Result<Option<Arc<Dentry_>>> {
+        match self.dentries.get(name) {
+            Some(Some(child)) => Ok(Some(child.clone())),
+            Some(None) => return_errno_with_message!(Errno::ENOENT, "found a negative dentry"),
+            None => Ok(None),
+        }
     }
 
+    /// Inserts a valid dentry if it's cacheable.
     pub fn insert(&mut self, name: String, dentry: Arc<Dentry_>) {
         // Do not cache it in the children if is not cacheable.
         // When we lookup it from the parent, it will always be newly created.
         if !dentry.inode.is_dentry_cacheable() {
             return;
```

Comment:
This check should be done by the caller. If the dentry is uncachable, then the caller would clone a string and/or a dentry for nothing. For example, the current usage pattern 

```rust
children.insert(String::from(new_name), dentry.clone());
```

should be deprecated in favor of the new one:

```rust
if dentry.inode.is_dentry_cacheable() {
    children.insert(String::from(new_name), dentry.clone());
}
```